### PR TITLE
Add update_by operation in Python client

### LINKED ISSUE #3496: Support the udpateBy Table operation in the Python client

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #3496

### CODE REVIEW

File: py/client/pydeephaven/query.py [hunk 1]
```diff
@@ -423,3 +423,15 @@ def agg_by(self, agg: ComboAggregation, by: List[str]):
             self
         """
         return super().agg_by(agg, by)
+
+    def update_by(self, ops: List[UpdateByOperation], by: List[str]):
```
[the remark this answers is not in the document]
Comment (chipkent) [reply py/client/pydeephaven/query.py:427]: does the client not support `by: Union(str,List[str])`?  If not, that is a deviation with the server that should be considered.

File: py/client/pydeephaven/updateby.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+GrpcUpdateByOperation = table_pb2.UpdateByRequest.UpdateByOperation
+GrpcUpdateByColumn = GrpcUpdateByOperation.UpdateByColumn
+GrpcUpdateBySpec = GrpcUpdateByColumn.UpdateBySpec
+GrpcUpdateByEma = GrpcUpdateBySpec.UpdateByEma
+GrpcUpdateByEmaOptions = GrpcUpdateByEma.UpdateByEmaOptions
+GrpcUpdateByEmaTimescale = GrpcUpdateByEma.UpdateByEmaTimescale
+GrpcUpdateByEmaTicks = GrpcUpdateByEmaTimescale.UpdateByEmaTicks
+GrpcUpdateByEmaTime = GrpcUpdateByEmaTimescale.UpdateByEmaTime
+GrpcMathContext = table_pb2.MathContext
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:19]: 1) Do these need to be public?
2) If they do need to be public, they should be documented.

File: py/client/pydeephaven/updateby.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class UpdateByBase(ABC):
+    @abstractmethod
+    def make_grpc_request(self):
+        ...
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:25]: private or doc

File: py/client/pydeephaven/updateby.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class MathContext(Enum):
+    """An Enum for predefined precision and rounding settings in numeric calculation."""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:29]:
```suggestion
    """An Enum for predefined precision and rounding settings in numeric calculations."""
```

File: py/client/pydeephaven/updateby.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+    UNLIMITED = 0, table_pb2.MathContext.RoundingMode.HALF_UP
+    """unlimited precision arithmetic, rounding is half-up"""
+
+    DECIMAL32 = 7, table_pb2.MathContext.RoundingMode.HALF_EVEN
+    """a precision setting matching the IEEE 754R Decimal32 format, 7 digits, rounding is half-even"""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:35]:
```suggestion
    """a precision setting matching the IEEE 754R Decimal32 format (32 bits, 7 decimal digits), rounding is half-even"""
```
same change on the others

File: py/client/pydeephaven/updateby.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class BadDataBehavior(Enum):
+    """An Enum defining ways to handle invalid data during update-by operations."""
+
+    RESET = table_pb2.BadDataBehavior.RESET
+    """Reset the state for the bucket to None when invalid data is encountered"""
+
+    SKIP = table_pb2.BadDataBehavior.SKIP
+    """Skip and do not process the invalid data without changing state"""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:51]:
```suggestion
    """Skip and do not process the invalid data.  The state is not changed."""
```

File: py/client/pydeephaven/updateby.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+    DECIMAL128 = 34, table_pb2.MathContext.RoundingMode.HALF_EVEN
+    """a precision setting matching the IEEE 754R Decimal128 format, 34 digits, rounding is half-even"""
+
+
+class BadDataBehavior(Enum):
+    """An Enum defining ways to handle invalid data during update-by operations."""
+
+    RESET = table_pb2.BadDataBehavior.RESET
+    """Reset the state for the bucket to None when invalid data is encountered"""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:48]:
```suggestion
    """Reset the state for the bucket to None when invalid data is encountered."""
```

File: py/client/pydeephaven/updateby.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+    SKIP = table_pb2.BadDataBehavior.SKIP
+    """Skip and do not process the invalid data without changing state"""
+
+    THROW = table_pb2.BadDataBehavior.THROW
+    """Throw an exception and abort processing when bad data is encountered"""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:54]:
```suggestion
    """Throw an exception and abort processing when bad data is encountered."""
```

File: py/client/pydeephaven/updateby.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+    POISON = table_pb2.BadDataBehavior.POISON
+    """Allow the bad data to poison the result. This is only valid for use with NaN"""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:57]:
```suggestion
    """Allow the bad data to poison the result. This is only valid for use with NaN."""
```

File: py/client/pydeephaven/updateby.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class OperationControl(UpdateByBase):
+    """A OperationControl represents control parameters for performing operations with the table
+        UpdateByOperation."""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:62]:
```suggestion
    """Control parameters specifying how table UpdateByOperations are computed."""
```

File: py/client/pydeephaven/updateby.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+
+    def __init__(self, on_null: BadDataBehavior = BadDataBehavior.SKIP,
+                 on_nan: BadDataBehavior = BadDataBehavior.SKIP,
+                 big_value_context: MathContext = MathContext.DECIMAL128):
+        """Initializes an OperationControl for use with certain UpdateByOperation, such as EMAs.
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:67]:
```suggestion
        """Creates control parameters specifying how UpdateByOperations are computed.
```

File: py/client/pydeephaven/updateby.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        self.big_value_context = big_value_context
+
+    def make_grpc_request(self):
+        return GrpcUpdateByEmaOptions(on_null_value=self.on_null.value, on_nan_value=self.on_nan.value,
+                                      big_value_context=GrpcMathContext(precision=self.big_value_context.value[0],
+                                                                        rounding_mode=self.big_value_context.value[1]))
+
+
+class UpdateByOperation(UpdateByBase):
+    """A UpdateByOperation represents an operator for the Table update-by operation."""
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:86]: 1) the definition feels circular
2) `update-by` or `update_by`?

File: py/client/pydeephaven/updateby.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+    def __init__(self, ub_column):
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:88]: type hint

File: py/client/pydeephaven/updateby.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            on_null (BadDataBehavior): the behavior for when null values are encountered, default is SKIP
+            on_nan (BadDataBehavior): the behavior for when NaN values are encountered, default is SKIP
+            big_value_context (MathContext): the context to use when processing arbitrary precision numeric values,
+                default is DECIMAL128.
+        """
+        self.on_null = on_null
+        self.on_nan = on_nan
+        self.big_value_context = big_value_context
+
+    def make_grpc_request(self):
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:79]: type hint?
This also comes back to the question of if we want to expose the Grpc stuff at the top of the file.

File: py/client/pydeephaven/updateby.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class UpdateByOperation(UpdateByBase):
+    """A UpdateByOperation represents an operator for the Table update-by operation."""
+
+    def __init__(self, ub_column):
+        self.ub_column = ub_column
+
+    def make_grpc_request(self):
+        return GrpcUpdateByOperation(column=self.ub_column)
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:92]: type hint?
This also comes back to the question of if we want to expose the Grpc stuff at the top of the file.

File: py/client/pydeephaven/updateby.py [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+
+
+def cum_sum(cols: List[str]) -> UpdateByOperation:
+    """ Create a cumulative sum update-by operation.
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:96]: `update-by` or `update_by`.  I can see the logic in either.  What did we do in the server?  Probably `update-by`.  Just be consistent with the server.  Check all locations.

File: py/client/pydeephaven/updateby.py [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+        cols (List[str]): the columns to be operated on, can be a common name or an equal expression,
+                i.e. "col_a = col_b" for different column names; when empty, update_by performs the cumulative sum
+                operation on all the applicable columns.
+
+    Returns:
+        UpdateByOperation
+    """
+    ub_spec = GrpcUpdateBySpec(sum=GrpcUpdateBySpec.UpdateByCumulativeSum())
+    ub_column = GrpcUpdateByColumn(spec=ub_spec, match_pairs=cols)
+    return UpdateByOperation(ub_column=ub_column)
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:108]: `ub_column` sounds singular, but it looks like it is created with multiple columns.  What is the correct nomenclature?  Apply everywhere it is relevant.

File: py/client/pydeephaven/updateby.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+
+    Returns:
+        UpdateByOperation
+    """
+    ub_spec = GrpcUpdateBySpec(min=GrpcUpdateBySpec.UpdateByCumulativeMin())
+    ub_column = GrpcUpdateByColumn(spec=ub_spec, match_pairs=cols)
+    return UpdateByOperation(ub_column=ub_column)
+
+
+def cum_max(cols: List[str]) -> UpdateByOperation:
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:143]: Server supports the union types for all of these: https://deephaven.io/core/pydoc/code/deephaven.updateby.html#deephaven.updateby.cum_sum

File: py/client/pydeephaven/updateby.py (hunk 16 above)
Comment (chipkent) [reply py/client/pydeephaven/updateby.py:96]: Server docs use `UpdateByOperation`.  Any reason to use different wording?
https://deephaven.io/core/pydoc/code/deephaven.updateby.html#deephaven.updateby.cum_sum

File: py/client/pydeephaven/updateby.py [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+        return GrpcUpdateByOperation(column=self.ub_column)
+
+
+def cum_sum(cols: List[str]) -> UpdateByOperation:
+    """ Create a cumulative sum update-by operation.
+
+    Args:
+        cols (List[str]): the columns to be operated on, can be a common name or an equal expression,
+                i.e. "col_a = col_b" for different column names; when empty, update_by performs the cumulative sum
+                operation on all the applicable columns.
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:101]: wording is not the same as the server.  any reason to change?

File: py/client/pydeephaven/updateby.py
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:1]: I have a window with the server docs pulled up, and there is a lot of deviation I see between the client and server.  If there isn't a good reason for the deviation, make them consistent.  This comment will override many of my other comments on this file.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/client/pydeephaven/_table_interface.py [hunk 20]
```diff
@@ -559,3 +560,19 @@ def agg_by(self, agg: ComboAggregation, by: List[str]):
         """
         table_op = ComboAggOp(column_names=by, combo_aggregation=agg)
         return self.table_op_handler(table_op)
+
+    def update_by(self, ops: List[UpdateByOperation], by: List[str]):
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/client/pydeephaven/_table_interface.py:564]: To do that will introduce circular reference and imports.

File: py/client/pydeephaven/updateby.py [hunk 21]
```diff
@@ -57,7 +57,7 @@ class BadDataBehavior(Enum):
     """Allow the bad data to poison the result. This is only valid for use with NaN"""
 
 
-class OperationControl(UpdateByBase):
+class OperationControl(_UpdateByBase):
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:60]: Is sphinx ok with this, or should it just be public?

File: py/client/pydeephaven/updateby.py [hunk 22]
```diff
@@ -103,13 +103,13 @@ def cum_sum(cols: List[str]) -> UpdateByOperation:
...
+    ub_column = _GrpcUpdateByColumn(spec=ub_spec, match_pairs=cols)
     return UpdateByOperation(ub_column=ub_column)
 
 
 def cum_prod(cols: List[str]) -> UpdateByOperation:
-    """ Create a cumulative product update-by operation.
+    """Creates a cumulative product UpdateByOperation for the supplied column names.
 
     Args:
         cols (List[str]): the columns to be operated on, can be a common name or an equal expression,
```
Comment (chipkent) [inline py/client/pydeephaven/updateby.py:115]: Check the args for consistency with the server as well.
```
cols (Union[str, List[str]]) – the column(s) to be operated on, can include expressions to rename the output, i.e. “new_col = col”; when empty, update_by performing the cumulative product operation on all the applicable columns.
```

**Review by chipkent — CHANGES_REQUESTED** [review]

File: py/client/pydeephaven/updateby.py (hunk 21 above)
Comment (jmao-denver) [reply py/client/pydeephaven/updateby.py:60]: Yeah, sphinx works great.

**Review by chipkent — APPROVED** [review]

**Pull request merged**